src/plugins/ucx/ucx_backend.cpp

```diff
@@ -617,14 +670,427 @@ nixlUcxThreadEngine::getNotifs(notif_list_t &notif_list) {
+
+    static nixlUcxDedicatedThread *
+    getDedicatedThread() {
+        return (nixlUcxDedicatedThread *)tlsThread();
+    }
+
+    void
+    addRequest(nixlUcxChunkBackendH *handle) {
+        requests_.push_back(handle);
+    }
+
+protected:
+    void
+    run() override {
+        auto guard = asio::make_work_guard(io_);
+        NIXL_DEBUG << "dedicated " << *this << " running";
+
+        while (!io_.stopped()) {
+            if (!requests_.empty()) {
+                io_.poll_one();
+            } else {
+                NIXL_TRACE << "dedicated " << *this << " waiting for requests";
+                io_.run_one();
+            }
+
+            if (requests_.empty()) {
+                continue;
+            }
+
+            for (auto it = requests_.begin(); it != requests_.end();) {
+                nixl_status_t status = (*it)->status();
+                if (status != NIXL_IN_PROG) {
+                    NIXL_TRACE << "dedicated " << *this << " completing " << *(*it)
+                               << " with status: " << status;
+                    (*it)->complete(status);
+                    it = requests_.erase(it);
+                } else {
+                    ++it;
+                }
+            }
+        }
+
+        if (!requests_.empty()) {
+            NIXL_WARN << "dedicated " << *this << " dropping " << requests_.size()
+                      << " requests on exit";
+            for (auto it = requests_.begin(); it != requests_.end();) {
+                NIXL_INFO << "dropping " << *(*it);
+                (*it)->complete(NIXL_ERR_BACKEND);
+            }
+            requests_.clear();
+        }
+
+        NIXL_DEBUG << "dedicated " << *this << " exiting";
+    }
+
+private:
+    asio::io_context &io_;
+    std::vector<nixlUcxChunkBackendH *> requests_;
+};
+
+nixlUcxThreadPoolEngine::nixlUcxThreadPoolEngine(const nixlBackendInitParams &init_params)
+    : nixlUcxEngine(init_params) {
+    size_t num_threads = nixl_b_params_get(init_params.customParams, "num_threads", 0);
+    numSharedWorkers_ = getWorkers().size() - num_threads;
+    NIXL_ASSERT(numSharedWorkers_ > 0);
+
+    splitBatchSize_ = nixl_b_params_get(init_params.customParams, "split_batch_size", 1024);
+
+    auto init = [this]() { nixlUcxEngine::vramApplyCtx(); };
+
+    if (init_params.enableProgTh) {
+        sharedThread_ = std::make_unique<nixlUcxSharedThread>(
+            this, init, numSharedWorkers_, init_params.pthrDelay);
+        for (size_t i = 0; i < numSharedWorkers_; i++) {
+            sharedThread_->addWorker(getWorkers()[i].get(), i);
+        }
+        sharedThread_->start();
+    }
+
+    if (num_threads > 0) {
+        io_.reset(new asio::io_context());
+        dedicatedThreads_.reserve(num_threads);
+        for (size_t i = 0; i < num_threads; ++i) {
+            size_t worker_id = numSharedWorkers_ + i;
+            dedicatedThreads_.emplace_back(
+                std::make_unique<nixlUcxDedicatedThread>(this, init, *io_));
```

Comment:
The fact that we construct another object in the object constructor and pass `this` as one of the parameters suggests that there might be a problem.
IIUC the only reason to pass `this` is to check that `nixlUcxThreadEngine::appendNotif` is invoked in progress thread.

That's not an issue of this PR, but something that can be simplified in a separate PR.